metadata_writer_nne.c

```diff
@@ -94,132 +309,725 @@ static uint8_t md_nne_handle_gps_event(struct md_writer_nne *mwn,
+    else
+        md_nne_add_json_key_value_int(obj, "mccmnc", msg->mccmnc);
+
+    md_nne_add_json_key_value_string(obj, "key", msg->key);
+    md_nne_add_json_key_value(obj, "value", msg->value);
+
+    if (msg->extra == NULL)
+        md_nne_add_json_key_value_null(obj, "extra");
+    else
+        md_nne_add_json_key_value_string(obj, "extra", msg->extra);
+
+    md_nne_add_json_key_value_string(obj, "source", NNE_MESSAGE_SOURCE_STR[msg->source]);
+
+    if (msg->type == NNE_MESSAGE_TYPE_BINS1MIN)
+        md_nne_add_json_key_value_int(obj, "delta", msg->delta);
+
+    META_PRINT_SYSLOG(mwn->parent, LOG_ERR, "NNE writer: mccmnc %d: %s\n", msg->mccmnc,
+                      json_object_to_json_string_ext(obj, JSON_C_TO_STRING_PLAIN));
+
+    if (mwn->metadata_cache == NULL)
+        mwn->metadata_cache = json_object_new_array();
+
+    json_object_array_add(mwn->metadata_cache, obj);
+}
+
+
+/*static void md_nne_process_mode(struct md_writer_nne *mwn,
+                                       struct nne_metadata_descr* descr,
+                                       struct nne_modem* modem,
+                                       struct md_iface_event *mie,
+                                       enum nne_message_source source)
+{
+
+}*/
+
+
+static void md_nne_process_iface_event(struct md_writer_nne *mwn,
+                                       struct nne_metadata_descr* descr,
+                                       struct nne_modem* modem,
+                                       struct md_iface_event *mie,
+                                       enum nne_message_source source)
+{
+    int i;
+    struct nne_value value =  descr->parse_cb(modem, mie);
+
+    if (nne_value_compare(modem->metadata[descr->idx].value, value) != 0) {
+
+        if (modem->metadata[descr->idx].value.type == NNE_TYPE_STRING)
+            free(modem->metadata[descr->idx].value.u.v_str);
+        modem->metadata[descr->idx].value = value;
+
+        META_PRINT_SYSLOG(mwn->parent, LOG_ERR, "NNE writer: mccmnc %d: %d,%s changed %d\n",
+                          mie->imsi_mccmnc, descr->idx, descr->key,
+                          modem->metadata[descr->idx].value.u.v_uint8);
+
+        // Generate json event
+        struct nne_message msg;
+        msg.type = NNE_MESSAGE_TYPE_EVENT;
+        msg.tstamp = mie->tstamp;
+        msg.node = mwn->node_id;
+        msg.mccmnc = mie->imsi_mccmnc;
+        msg.key = descr->key;
+        msg.value = value;
+        msg.extra = NULL;
+        msg.source = source;
+        msg.delta = 0;
+
+        md_nne_send_message(mwn, &msg);
+
+        // if mode has changed nullify mode dependend metadata
+        if (descr->idx == NNE_IDX_MODE) {
+            for(i = 0; i <= NNE_IDX_MAX; i++) {
+                if (descr->mode_dependent) {
+                    if (modem->metadata[i].value.type == NNE_TYPE_STRING) {
+                        free(modem->metadata[i].value.u.v_str);
+                    }
```

Comment:
You seem to do the same thing three times (freeing the memory of the string), can this be merged somehow?